engine/apps/alerts/models/alert_receive_channel.py

```diff
@@ -206,6 +204,17 @@ class AlertReceiveChannel(IntegrationOptionsMixin, MaintainableObject):
     rate_limited_in_slack_at = models.DateTimeField(null=True, default=None)
     rate_limit_message_task_id = models.CharField(max_length=100, null=True, default=None)
 
+    AlertGroupCustomLabels = list[tuple[str, str | None, str | None]]
+    alert_group_labels_custom: AlertGroupCustomLabels = models.JSONField(null=True, default=list)
+    """
+    Stores "custom labels" for alert group labels. Custom labels can be either "plain" or "templated".
+    For plain labels, the format is: [<LABEL_KEY_ID>, <LABEL_VALUE_ID>, None]
+    For templated labels, the format is: [<LABEL_KEY_ID>, None, <JINJA2_TEMPLATE>]
+    """
```

Comment:
Not blocking, but wondering if you have considered a dict instead? (`{<LABEL_KEY_ID>: {"value_id": None, "template": <JINJA2_TEMPLATE>}, ...}`) It may be easier to read in the code, and simpler to extend or change later?

Reply:
We want to keep these custom labels ordered, so I considered doing this:

```python3
class AlertGroupCustomLabel(typing.TypedDict):
    key_id: str
    value_id: str | None
    template: str | None

AlertGroupCustomLabels = list[AlertGroupCustomLabel]
```

It looked too verbose to me (making list comprehensions like [this one](https://github.com/grafana/oncall/pull/3385/files#diff-b8990ca4b389382d4e7b6f58fd7c1bf0773c85d3a86db2a4a4486e2459dcf8ffR186-R199) much harder to read) + takes more space in the DB, so I went with the tuple approach. Let's try this out and see if we have any issues upgrading the DB schema later on? 🙂

Reply:
Makes sense :+1: